src/main/scala/viper/gobra/translator/encodings/slices/SliceEncoding.scala

```diff
@@ -0,0 +1,461 @@
+      } yield vpr.CondExp(
+        vpr.EqCmp(expT, nilSlice(typT)(ctx)(pos, info, errT))(pos, info, errT),
+        vpr.IntLit(0)(pos, info, errT),
+        ctx.slice.cap(ctx.option.get(expT, ctx.slice.typ(typT))(pos, info, errT))(pos, info, errT)
+      )(pos, info, errT)
+
+      case exp @ in.Slice((base : in.Location) :: ctx.Array(_, _) / Shared, low, high, max) => for {
+        baseT <- ctx.typeEncoding.reference(ctx)(base)
+        lowT <- goE(low)
+        highT <- goE(high)
+        maxOptT <- option(max match {
+          case Some(e) => Some(goE(e))
+          case None => None
+        })
+      } yield {
+        val (pos, info, errT) = exp.vprMeta
+        val unboxedBaseT = arrayEmb.unbox(baseT, base.typ.asInstanceOf[in.ArrayT])(base)(ctx)
+
+        ctx.option.some(maxOptT match {
+          case None => sliceFromArray(vpr.Ref, unboxedBaseT, lowT, highT)(ctx)(pos, info, errT)
+          case Some(maxT) => fullSliceFromArray(vpr.Ref, unboxedBaseT, lowT, highT, maxT)(ctx)(pos, info, errT)
+        })(pos, info, errT)
+      }
+
+      case exp @ in.Slice((base : in.Expr) :: ctx.Slice(typ), low, high, max) => for {
+        baseT <- goE(base)
+        lowT <- goE(low)
+        highT <- goE(high)
+        maxOptT <- option(max match {
+          case Some(e) => Some(goE(e))
+          case None => None
+        })
+      } yield {
+        val (pos, info, errT) = exp.vprMeta
+        val typT = ctx.typeEncoding.typ(ctx)(typ)
+        val baseOptT = ctx.option.get(baseT, ctx.slice.typ(typT))(pos, info, errT)
+
+        ctx.option.some(maxOptT match {
+          case None => sliceFromSlice(vpr.Ref, baseOptT, lowT, highT)(ctx)(pos, info, errT)
+          case Some(maxT) => fullSliceFromSlice(vpr.Ref, baseOptT, lowT, highT, maxT)(ctx)(pos, info, errT)
+        })(pos, info, errT)
+      }
+
+      case (lit : in.SliceLit) :: ctx.Slice(_) => {
+        val litA = lit.asArrayLit
+        val tmp = in.LocalVar(Names.freshName, litA.typ.withAddressability(Addressability.pointerBase))(lit.info)
+        val tmpT = ctx.typeEncoding.variable(ctx)(tmp)
+        for {
+          initT <- ctx.typeEncoding.initialization(ctx)(tmp)
+          assignT <- ctx.typeEncoding.assignment(ctx)(in.Assignee.Var(tmp), litA, lit)
+          sliceT <- ctx.expr.translate(in.Slice(tmp, in.IntLit(0)(lit.info), in.IntLit(litA.length)(lit.info), None)(lit.info))(ctx)
+          _ <- local(tmpT)
+          _ <- write(initT)
+          _ <- write(assignT)
+        } yield sliceT
+      }
+    }
+  }
+
+  override def reference(ctx : Context) : in.Location ==> CodeWriter[vpr.Exp] = default(super.reference(ctx)) {
+    case (exp @ in.IndexedExp(base :: ctx.Slice(typ), idx)) :: _ / Shared => for {
+      baseT <- ctx.expr.translate(base)(ctx)
+      idxT <- ctx.expr.translate(idx)(ctx)
+      typT = ctx.typeEncoding.typ(ctx)(typ)
+    } yield {
+      val (pos, info, errT) = exp.vprMeta
+      val getT = ctx.option.get(baseT, ctx.slice.typ(typT))(pos, info, errT)
```

Comment:
What happens if the slice is nil? Also, what happens with out of bounds access?

Reply:
A verification error with a suitable message will now be presented in case of an out-of-bounds access.

Nil slices should be handled correctly, as they are encoded roughly as a default array that is sliced to have capacity `0`.